diff --git a/src/solver/include/epanetmsx.h b/src/solver/include/epanetmsx.h
index 285217f..4d1fd06 100644
--- a/src/solver/include/epanetmsx.h
+++ b/src/solver/include/epanetmsx.h
@@ -5,11 +5,11 @@
 **  AUTHORS:       see AUTHORS
 **  Copyright:     see AUTHORS
 **  License:       see LICENSE
-**  VERSION:       2.0.0
+**  VERSION:       2.0.0
 **  LAST UPDATE:   11/01/10
 *******************************************************************************/

-#include "epanet2.h"                   // EPANET toolkit header file
+#include "epanet2_2.h"                   // EPANET toolkit header file

 #ifndef EPANETMSX_H
 #define EPANETMSX_H
@@ -55,9 +55,7 @@ extern "C" {

 // --- declare MSX functions

-int  MSXDLLEXPORT MSXENopen(const char *inpFile, const char *rptFile,
-                 const char *outFile);
-int  MSXDLLEXPORT MSXopen(char *fname);
+int  MSXDLLEXPORT MSXopen(EN_Project ph, char *fname);
 int  MSXDLLEXPORT MSXsolveH(void);
 int  MSXDLLEXPORT MSXusehydfile(char *fname);
 int  MSXDLLEXPORT MSXsolveQ(void);
@@ -67,7 +65,6 @@ int  MSXDLLEXPORT MSXsaveoutfile(char *fname);
 int  MSXDLLEXPORT MSXsavemsxfile(char *fname);
 int  MSXDLLEXPORT MSXreport(void);
 int  MSXDLLEXPORT MSXclose(void);
-int  MSXDLLEXPORT MSXENclose(void);

 int  MSXDLLEXPORT MSXgetindex(int type, char *id, int *index);
 int  MSXDLLEXPORT MSXgetIDlen(int type, int index, int *len);
diff --git a/src/solver/msxerr.c b/src/solver/msxerr.c
index 63dfa06..c24db33 100644
--- a/src/solver/msxerr.c
+++ b/src/solver/msxerr.c
@@ -13,7 +13,7 @@
 #include <string.h>

 #include "msxtypes.h"
-#include "epanet2.h"
+#include "epanet2_2.h"

 //  External variables
 //--------------------
@@ -67,8 +67,8 @@ void MSXerr_writeMathErrorMsg()
 **    writes math error message to EPANET report file.
 */
 {
-	ENwriteline(mathErrorMsg);
-	ENwriteline("");
+	EN_writeline(MSX.ph, mathErrorMsg);
+	EN_writeline(MSX.ph, "");
 }

 //=============================================================================
diff --git a/src/solver/msxfile.c b/src/solver/msxfile.c
index 26cc66d..cbb6f08 100644
--- a/src/solver/msxfile.c
+++ b/src/solver/msxfile.c
@@ -15,7 +15,7 @@
 #include "msxtypes.h"
 #include "msxutils.h"
 #include "msxdict.h"
-#include "epanet2.h"
+#include "epanet2_2.h"

 //  External variables
 //--------------------
@@ -152,7 +152,7 @@ void  saveParams(FILE *f)
             {
                 if ( MSX.Link[i].param[j] != x )
                 {
-                    ENgetlinkid(i, id);
+                    EN_getlinkid(MSX.ph, i, id);
                     fprintf(f, "\nPIPE    %-32s  %-32s  %e",
                         id, MSX.Param[j].id, MSX.Link[i].param[j]);
                 }
@@ -162,7 +162,7 @@ void  saveParams(FILE *f)
                 if ( MSX.Tank[i].param[j] != x )
                 {
                     k = MSX.Tank[i].node;
-                    ENgetnodeid(k, id);
+                    EN_getnodeid(MSX.ph, k, id);
                     fprintf(f, "\nTANK    %-32s  %-32s  %e",
                         id, MSX.Param[j].id, MSX.Tank[i].param[j]);
                 }
@@ -189,7 +189,7 @@ void  saveQuality(FILE *f)
         {
             if ( MSX.Node[i].c0[j] > 0.0 && MSX.Node[i].c0[j] != MSX.C0[j])
             {
-                ENgetnodeid(i, id);
+                EN_getnodeid(MSX.ph, i, id);
                 fprintf(f, "\nNODE    %-32s  %-32s  %e",
                     id, MSX.Species[j].id, MSX.Node[i].c0[j]);
             }
@@ -198,7 +198,7 @@ void  saveQuality(FILE *f)
         {
             if ( MSX.Link[i].c0[j] > 0.0 && MSX.Link[i].c0[j] != MSX.C0[j])
             {
-                ENgetlinkid(i, id);
+                EN_getlinkid(MSX.ph, i, id);
                 fprintf(f, "\nLINK    %-32s  %-32s  %e",
                     id, MSX.Species[j].id, MSX.Link[i].c0[j]);
             }
@@ -222,7 +222,7 @@ void  saveSources(FILE *f)
         {
             if ( source->c0 > 0.0 && source->type > -1)   //Feng Shang 09/23/2008
             {
-                ENgetnodeid(i, id);
+                EN_getnodeid(MSX.ph, i, id);
                 fprintf(f, "\n%-10s  %-32s  %-32s  %e",
                     SourceTypeWords[source->type], id,
                     MSX.Species[source->species].id, source->c0);
diff --git a/src/solver/msxinp.c b/src/solver/msxinp.c
index cebf547..171cc48 100644
--- a/src/solver/msxinp.c
+++ b/src/solver/msxinp.c
@@ -18,7 +18,7 @@
 #include "msxtypes.h"
 #include "msxutils.h"
 #include "msxdict.h"
-#include "epanet2.h"
+#include "epanet2_2.h"

 //  Constants
 //-----------
@@ -133,8 +133,8 @@ int MSXinp_countMsxObjects()
     strcpy(MSX.Msg, "Processing MSX input file ");
     strcpy(line, MSX.MsxFile.name);
     strcat(MSX.Msg, line);
-    ENwriteline(MSX.Msg);
-    ENwriteline("");
+    EN_writeline(MSX.ph, MSX.Msg);
+    EN_writeline(MSX.ph, "");

 // --- make pass through EPANET-MSX data file counting number of each object

@@ -192,9 +192,9 @@ int  MSXinp_countNetObjects()

 // --- retrieve number of network elements

-    CALL(errcode, ENgetcount(EN_NODECOUNT, &MSX.Nobjects[NODE]));
-    CALL(errcode, ENgetcount(EN_TANKCOUNT, &MSX.Nobjects[TANK]));
-    CALL(errcode, ENgetcount(EN_LINKCOUNT, &MSX.Nobjects[LINK]));
+    CALL(errcode, EN_getcount(MSX.ph, EN_NODECOUNT, &MSX.Nobjects[NODE]));
+    CALL(errcode, EN_getcount(MSX.ph, EN_TANKCOUNT, &MSX.Nobjects[TANK]));
+    CALL(errcode, EN_getcount(MSX.ph, EN_LINKCOUNT, &MSX.Nobjects[LINK]));
     return errcode;
 }

@@ -216,22 +216,22 @@ int MSXinp_readNetData()
     int   i, k, n, t = 0;
     int   n1 = 0, n2 = 0;
     long  qstep;
-    float diam = 0.0, len = 0.0, v0 = 0.0, xmix = 0.0, vmix = 0.0;
+    double diam = 0.0, len = 0.0, v0 = 0.0, xmix = 0.0, vmix = 0.0;

-    float roughness = 0.0;
+    double roughness = 0.0;

 // --- get flow units & time parameters

-    CALL(errcode, ENgetflowunits(&MSX.Flowflag));
+    CALL(errcode, EN_getflowunits(MSX.ph, &MSX.Flowflag));
     if ( MSX.Flowflag >= EN_LPS ) MSX.Unitsflag = SI;
     else                          MSX.Unitsflag = US;
-    CALL(errcode, ENgettimeparam(EN_QUALSTEP, &qstep));
+    CALL(errcode, EN_gettimeparam(MSX.ph, EN_QUALSTEP, &qstep));
     MSX.Qstep = qstep * 1000;
-    CALL(errcode, ENgettimeparam(EN_REPORTSTEP, &MSX.Rstep));
-    CALL(errcode, ENgettimeparam(EN_REPORTSTART, &MSX.Rstart));
-    CALL(errcode, ENgettimeparam(EN_PATTERNSTEP, &MSX.Pstep));
-    CALL(errcode, ENgettimeparam(EN_PATTERNSTART, &MSX.Pstart));
-    CALL(errcode, ENgettimeparam(EN_STATISTIC, &MSX.Statflag));
+    CALL(errcode, EN_gettimeparam(MSX.ph, EN_REPORTSTEP, &MSX.Rstep));
+    CALL(errcode, EN_gettimeparam(MSX.ph, EN_REPORTSTART, &MSX.Rstart));
+    CALL(errcode, EN_gettimeparam(MSX.ph, EN_PATTERNSTEP, &MSX.Pstep));
+    CALL(errcode, EN_gettimeparam(MSX.ph, EN_PATTERNSTART, &MSX.Pstart));
+    CALL(errcode, EN_gettimeparam(MSX.ph, EN_STATISTIC, &MSX.Statflag));

 // --- read tank/reservoir data

@@ -241,10 +241,10 @@ int MSXinp_readNetData()
         k = i - n;
         if ( k > 0 )
         {
-            CALL(errcode, ENgetnodetype(i, &t));
-            CALL(errcode, ENgetnodevalue(i, EN_INITVOLUME, &v0));
-            CALL(errcode, ENgetnodevalue(i, EN_MIXMODEL,   &xmix));
-            CALL(errcode, ENgetnodevalue(i, EN_MIXZONEVOL, &vmix));
+            CALL(errcode, EN_getnodetype(MSX.ph, i, &t));
+            CALL(errcode, EN_getnodevalue(MSX.ph, i, EN_INITVOLUME, &v0));
+            CALL(errcode, EN_getnodevalue(MSX.ph, i, EN_MIXMODEL,   &xmix));
+            CALL(errcode, EN_getnodevalue(MSX.ph, i, EN_MIXZONEVOL, &vmix));
             if ( !errcode )
             {
                 MSX.Node[i].tank = k;
@@ -262,10 +262,10 @@ int MSXinp_readNetData()

     for (i=1; i<=MSX.Nobjects[LINK]; i++)
     {
-        CALL(errcode, ENgetlinknodes(i, &n1, &n2));
-        CALL(errcode, ENgetlinkvalue(i, EN_DIAMETER, &diam));
-        CALL(errcode, ENgetlinkvalue(i, EN_LENGTH, &len));
-        CALL(errcode, ENgetlinkvalue(i, EN_ROUGHNESS, &roughness));
+        CALL(errcode, EN_getlinknodes(MSX.ph, i, &n1, &n2));
+        CALL(errcode, EN_getlinkvalue(MSX.ph, i, EN_DIAMETER, &diam));
+        CALL(errcode, EN_getlinkvalue(MSX.ph, i, EN_LENGTH, &len));
+        CALL(errcode, EN_getlinkvalue(MSX.ph, i, EN_ROUGHNESS, &roughness));
         if ( !errcode )
         {
             MSX.Link[i].n1 = n1;
@@ -1009,7 +1009,7 @@ int parseQuality()

     else if ( i == 2 )
     {
-        err = ENgetnodeindex(Tok[1], &j);
+        err = EN_getnodeindex(MSX.ph, Tok[1], &j);
         if ( err ) return ERR_NAME;
         if ( MSX.Species[m].type == BULK ) MSX.Node[j].c0[m] = x;
     }
@@ -1018,7 +1018,7 @@ int parseQuality()

     else if ( i == 3 )
     {
-        err = ENgetlinkindex(Tok[1], &j);
+        err = EN_getlinkindex(MSX.ph, Tok[1], &j);
         if ( err ) return ERR_NAME;
         MSX.Link[j].c0[m] = x;
     }
@@ -1055,7 +1055,7 @@ int parseParameter()

     if ( MSXutils_match(Tok[0], "PIPE") )
     {
-        err = ENgetlinkindex(Tok[1], &j);
+        err = EN_getlinkindex(MSX.ph, Tok[1], &j);
         if ( err ) return ERR_NAME;
         MSX.Link[j].param[i] = x;
     }
@@ -1064,7 +1064,7 @@ int parseParameter()

     else if ( MSXutils_match(Tok[0], "TANK") )
     {
-        err = ENgetnodeindex(Tok[1], &j);
+        err = EN_getnodeindex(MSX.ph, Tok[1], &j);
         if ( err ) return ERR_NAME;
         j = MSX.Node[j].tank;
         if ( j > 0 ) MSX.Tank[j].param[i] = x;
@@ -1099,7 +1099,7 @@ int parseSource()

 // --- get node index

-    err = ENgetnodeindex(Tok[1], &j);
+    err = EN_getnodeindex(MSX.ph, Tok[1], &j);
     if ( err ) return ERR_NAME;

 //  --- get species index
@@ -1230,7 +1230,7 @@ int parseReport()
         }
         else for (i=1; i<Ntokens; i++)
         {
-            err = ENgetnodeindex(Tok[i], &j);
+            err = EN_getnodeindex(MSX.ph, Tok[i], &j);
             if ( err ) return ERR_NAME;
             MSX.Node[j].rpt = 1;
         }
@@ -1248,7 +1248,7 @@ int parseReport()
         }
         else for (i=1; i<Ntokens; i++)
         {
-            err = ENgetlinkindex(Tok[i], &j);
+            err = EN_getlinkindex(MSX.ph, Tok[i], &j);
             if ( err ) return ERR_NAME;
             MSX.Link[j].rpt = 1;
         }
@@ -1436,9 +1436,9 @@ void writeInpErrMsg(int errcode, char *sect, char *line, int lineCount)
         sprintf(msg, "%s at line %d of %s] section:",
              InpErrorTxt[errcode-INP_ERR_FIRST], lineCount, sect);
     }
-    ENwriteline("");
-    ENwriteline(msg);
-    ENwriteline(line);
+    EN_writeline(MSX.ph, "");
+    EN_writeline(MSX.ph, msg);
+    EN_writeline(MSX.ph, line);
 }

 //=============================================================================
@@ -1467,7 +1467,7 @@ int checkCyclicTerms()
 	{
             sprintf(msg, "Error 410 - term %s contains a cyclic reference.",
                 MSX.Term[i].id);
-            ENwriteline(msg);
+            EN_writeline(MSX.ph, msg);
             return 1;
         }
     }
diff --git a/src/solver/msxout.c b/src/solver/msxout.c
index 2bf7441..8b966b1 100644
--- a/src/solver/msxout.c
+++ b/src/solver/msxout.c
@@ -1,3 +1,4 @@
+#include <stdlib.h>
 /******************************************************************************
 **  MODULE:        MSXOUT.C
 **  PROJECT:       EPANET-MSX
diff --git a/src/solver/msxproj.c b/src/solver/msxproj.c
index c9c45ac..3ee26cc 100644
--- a/src/solver/msxproj.c
+++ b/src/solver/msxproj.c
@@ -20,7 +20,7 @@
 //#include "mempool.h"
 #include "hash.h"
 #include "smatrix.h"
-#include "epanet2.h"
+#include "epanet2_2.h"
 #include "dispersion.h"
 //  Exported variables
 //--------------------
@@ -141,8 +141,8 @@ int  MSXproj_open(char *fname)

     if (MSX.DispersionFlag != 0)
     {
-        float relvis;
-        ENgetoption(13, &relvis);
+        double relvis;
+        EN_getoption(MSX.ph, 13, &relvis);
         MSX.Dispersion.viscosity = relvis * 1.1E-5;

         msx_createsparse();   //symmetric matrix
diff --git a/src/solver/msxrpt.c b/src/solver/msxrpt.c
index c71a3e1..2aedf9b 100644
--- a/src/solver/msxrpt.c
+++ b/src/solver/msxrpt.c
@@ -16,7 +16,7 @@
 #include <math.h>

 #include "msxtypes.h"
-#include "epanet2.h"
+#include "epanet2_2.h"

 // Constants
 //----------
@@ -134,7 +134,7 @@ void createSeriesTables()
     for (j=1; j<=MSX.Nobjects[NODE]; j++)
     {
         if ( !MSX.Node[j].rpt ) continue;
-        ENgetnodeid(j, IDname);
+        EN_getnodeid(MSX.ph, j, IDname);
         createTableHdr(NODE, SERIES_TABLE);
         writeNodeTable(j, SERIES_TABLE);
     }
@@ -144,7 +144,7 @@ void createSeriesTables()
     for (j=1; j<=MSX.Nobjects[LINK]; j++)
     {
         if ( !MSX.Link[j].rpt ) continue;
-        ENgetlinkid(j, IDname);
+        EN_getlinkid(MSX.ph, j, IDname);
         createTableHdr(LINK, SERIES_TABLE);
         writeLinkTable(j, SERIES_TABLE);
     }
@@ -261,7 +261,7 @@ void  writeNodeTable(int j, int tableType)
         }
         if ( tableType == STATS_TABLE )
         {
-            ENgetnodeid(j, IDname);
+            EN_getnodeid(MSX.ph, j, IDname);
             sprintf(Line, "%-16s", IDname);
         }
         for (m=1; m<=MSX.Nobjects[SPECIES]; m++)
@@ -293,7 +293,7 @@ void  writeLinkTable(int j, int tableType)
         }
         if ( tableType == STATS_TABLE )
         {
-            ENgetlinkid(j, IDname);
+            EN_getlinkid(MSX.ph, j, IDname);
             sprintf(Line, "%-16s", IDname);
         }
         for (m=1; m<=MSX.Nobjects[SPECIES]; m++)
@@ -341,7 +341,7 @@ void  writeLine(char *line)
 {
     if ( LineNum == MSX.PageSize ) newPage();
     if ( MSX.RptFile.file ) fprintf(MSX.RptFile.file, "  %s\n", line);   //(modified, FS-01/07/2008)
-    else ENwriteline(line);
+    else EN_writeline(MSX.ph, line);
     LineNum++;
 }

diff --git a/src/solver/msxtoolkit.c b/src/solver/msxtoolkit.c
index 51d6b03..3c3903e 100644
--- a/src/solver/msxtoolkit.c
+++ b/src/solver/msxtoolkit.c
@@ -25,8 +25,8 @@
 #include <stdlib.h>

 #include "msxtypes.h"
-#include "msxutils.h"
-#include "epanet2.h"
+#include "msxutils.h"
+#include "epanet2_2.h"
 #include "epanetmsx.h"

 //  External variables
@@ -52,39 +52,7 @@ int    MSXfile_save(FILE *f);

 //=============================================================================

-int MSXDLLEXPORT   MSXENopen(const char *inpFile, const char *rptFile, const char *outFile)
-/*
-**  Purpose:
-**    pass-thru to open the EPANET toolkit system
-**
-**  Input:
-**    inpFile = name of the EPANET input file
-**    rptFile = name of the EPANET report file
-**    outFile = name of the EPANET binary file
-**
-**  Returns:
-**    an error code (or 0 for no error);
-*/
-{
-    int err = 0;
-    err = ENopen(inpFile, rptFile, outFile);
-    return err;
-}
-
-int MSXDLLEXPORT   MSXENclose(void)
-/*
-**  Purpose:
-**    pass-thru to open the EPANET toolkit system
-**
-*/
-{
-    int err = 0;
-    err = ENclose();
-    return err;
-}
-
-
-int  MSXDLLEXPORT  MSXopen(char *fname)
+int  MSXDLLEXPORT  MSXopen(EN_Project ph, char *fname)
 /*
 **  Purpose:
 **    opens the EPANET-MSX toolkit system.
@@ -98,13 +66,14 @@ int  MSXDLLEXPORT  MSXopen(char *fname)
 {
     int err = 0;
     if (MSX.ProjectOpened) return(ERR_MSX_OPENED);
+    MSX.ph = ph;
     CALL(err, MSXproj_open(fname));
     CALL(err, MSXqual_open());

     if ( err )
     {
-        ENwriteline(MSXproj_getErrmsg(err));
-        ENwriteline("");
+        EN_writeline(MSX.ph, MSXproj_getErrmsg(err));
+        EN_writeline(MSX.ph, "");
     }

     return err;
@@ -146,8 +115,8 @@ int   MSXDLLEXPORT  MSXsolveH()

 // --- use EPANET to solve for & save hydraulics results

-    CALL(err, ENsolveH());
-    CALL(err, ENsavehydfile(MSX.HydFile.name));
+    CALL(err, EN_solveH(MSX.ph));
+    CALL(err, EN_savehydfile(MSX.ph, MSX.HydFile.name));
     CALL(err, MSXusehydfile(MSX.HydFile.name));
     return err;
 }
diff --git a/src/solver/msxtypes.h b/src/solver/msxtypes.h
index 836ad5d..526c13f 100644
--- a/src/solver/msxtypes.h
+++ b/src/solver/msxtypes.h
@@ -17,6 +17,7 @@

 #include "mathexpr.h"
 #include "mempool.h"
+#include "epanet2_2.h"

 //-----------------------------------------------------------------------------
 //  Definition of 4-byte integers & reals
@@ -467,6 +468,8 @@ typedef struct

 typedef struct                         // MSX PROJECT VARIABLES
 {
+   EN_Project ph;                      // EPANET 2.2 project handle
+
    TFile  HydFile,                     // EPANET hydraulics file
           MsxFile,                     // MSX input file
           OutFile,                     // MSX binary output file
